servicetalk-transport-netty-internal/src/test/java/io/servicetalk/transport/netty/internal/NonPipelinedCloseHandlerTest.java

```diff
@@ -0,0 +1,273 @@
+    private final AtomicBoolean closed = new AtomicBoolean();
+    private final List<Events> events;
+    private final ExpectEvent expectEvent;
+    @Nullable
+    private CloseEvent observedEvent;
+
+    protected enum Events {
+        IB, IE, IC, // emit Input Begin/End/Closing
+        OB, OE, OC, // emit Output Begin/End/Closing
+        IS, OS,     // emit Input/Output Socket closed
+        UC,         // emit User Closing
+        FC,         // Full-Close
+        CI, CO,     // emit Inbound/Outbound close (read cancellation, write subscriber termination)
+    }
+
+    protected enum ExpectEvent {
+        NIL(null), // No event, not closed
+        PCO(PROTOCOL_CLOSING_OUTBOUND),
+        PCI(PROTOCOL_CLOSING_INBOUND),
+        GUC(GRACEFUL_USER_CLOSING),
+        CCO(CHANNEL_CLOSED_OUTBOUND),
+        CCI(CHANNEL_CLOSED_INBOUND);
+
+        @Nullable
+        private final CloseEvent ce;
+
+        ExpectEvent(@Nullable CloseEvent c) {
+            this.ce = c;
+        }
+    }
+
+    public NonPipelinedCloseHandlerTest(boolean client, final List<Events> events, final ExpectEvent expectEvent,
+                                        final String desc, final String location) {
+        h = new NonPipelinedCloseHandler(client);
+        this.events = events;
+        this.expectEvent = expectEvent;
+        ctx = mock(ChannelHandlerContext.class);
+        channel = mock(SocketChannel.class, "[id: 0xmocked, L:mocked - R:mocked]");
+        when(ctx.channel()).thenReturn(channel);
+        when(ctx.newPromise()).thenReturn(new DefaultChannelPromise(channel));
+        when(channel.newPromise()).thenReturn(new DefaultChannelPromise(channel));
+
+        EventExecutor exec = mock(EventExecutor.class);
+        when(ctx.executor()).thenReturn(exec);
+        when(exec.inEventLoop()).thenReturn(true);
+        EventLoop loop = mock(EventLoop.class);
+        when(channel.eventLoop()).thenReturn(loop);
+        when(loop.inEventLoop()).thenReturn(true);
+        ChannelPipeline pipeline = mock(ChannelPipeline.class);
+        when(ctx.pipeline()).thenReturn(pipeline);
+        when(channel.pipeline()).thenReturn(pipeline);
+        ChannelFuture future = mock(ChannelFuture.class);
+        when(channel.close()).then(__ -> {
+            closed.set(true);
+            return future;
+        });
+        h.registerEventHandler(channel, e -> {
+            if (observedEvent == null) {
+                observedEvent = e;
+            }
+        });
+    }
+
+    @Parameterized.Parameters(name = "{index}. {3} - {0} {1} = {2}")
+    public static Collection<Object[]> data() {
+        StackTraceElement se = Thread.currentThread().getStackTrace()[1];
+        List<Object[]> params = asList(new Object[][] {
+                {true, e(OB, OE, IB, IE), NIL, "sequential, no close"},
+                {true, e(OB, IB, OE, IE), NIL, "interleaved, no close"},
+                {true, e(OB, CI, OE, FC), NIL, "inbound close while writing"},
```

Comment:
@idelpivnitskiy - I duplicated the tests for now bcz not all states/verification are applicable and `RequestResponseCloseHandlerTest`'s usage of `spy` means the class cannot be final.

Note that there are a bunch of test cases where the inbound/outbound close event occurs that we have an opportunity to enhance. lets chat offline.